### ISSUE TITLE
Does CLX support Compose key sequences?

### ISSUE THREAD

**Issue by eadmund** [sharplispers/clx#30]
I'm trying to type Compose key sequences in StumpWM, and see that they don't work — but it does recognise that it's pressed (as the SunCompose key). I _think_ that the right place to handle this is in CLX, not in xkeyboard.

I'm willing to take a cut at implementing this: I assume that EVENT-LOOP is the right place to get started. Does that sound right?

**Comment by fiddlerwoaroof** [comment]
I'm not sure if there has been any progress on this, but I've been looking into this in order to improve text-input support for McCLIM.  There are two halves to this problem: one is getting CLX to understand the correct keysym->character mappings.  This can be deduced from the file /usr/include/X11/keysymdefs.h (or similar) and I've collected all keysyms I could programmatically extract from that file in a gist:

https://gist.github.com/fiddlerwoaroof/5b62a4be6b115c08665d1d5989d3b4e7

The other part, which I still haven't looked into, is properly handling the compose key sequences.  When using xev to trace a usage of the compose key, one gets the following output. Note that the characters that are composed have XFilterEvent returning true whereas the character to be displayed has XFilterEvent be false.  So, I _think_ what needs to be done is detect the compose key event and then ignore the filtered events in favor of the final one.

```
KeyPress event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666845447, (338,1493), root:(1939,1513),
    state 0x0, keycode 108 (keysym 0xff20, Multi_key), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: True

KeyRelease event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666845599, (338,1493), root:(1939,1513),
    state 0x0, keycode 108 (keysym 0xff20, Multi_key), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666850775, (945,1549), root:(2546,1569),
    state 0x0, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666851039, (945,1549), root:(2546,1569),
    state 0x1, keycode 17 (keysym 0x2a, asterisk), same_screen YES,
    XLookupString gives 1 bytes: (2a) "*"
    XmbLookupString gives 1 bytes: (2a) "*"
    XFilterEvent returns: True

KeyRelease event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666851135, (945,1549), root:(2546,1569),
    state 0x1, keycode 17 (keysym 0x2a, asterisk), same_screen YES,
    XLookupString gives 1 bytes: (2a) "*"
    XFilterEvent returns: False

KeyRelease event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666851639, (945,1549), root:(2546,1569),
    state 0x1, keycode 50 (keysym 0xffe1, Shift_L), same_screen YES,
    XLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666852799, (945,1549), root:(2546,1569),
    state 0x0, keycode 38 (keysym 0x61, a), same_screen YES,
    XLookupString gives 1 bytes: (61) "a"
    XmbLookupString gives 1 bytes: (61) "a"
    XFilterEvent returns: True

KeyPress event, serial 36, synthetic NO, window 0x4200001,
    root 0x4d9, subw 0x0, time 666852799, (945,1549), root:(2546,1569),
    state 0x0, keycode 0 (keysym 0xe5, aring), same_screen YES,
    XLookupString gives 0 bytes:
    XmbLookupString gives 2 bytes: (c3 a5) "å"
    XFilterEvent returns: False
```

**Comment by equwal** [comment]
Good start! Here is some documentation:
https://www.x.org/releases/X11R7.7/doc/man/man5/Compose.5.xhtml

So ideally a program would be able to parse the compose DSL, and search for the first file in the order of precedence provided.

**Comment by JMC-design** [comment]
This should be external to CLX. It was in xlib because they were throwing in everything including the kitchen sink.  It is not present in xcb. Client libraries should learn how to detect compatability modes for groups and level shifts.

**Comment by eadmund** [comment]
A quick Googling indicates that the lack of an xcb equivalent to XLookupString _seems_ to be a case of feature-incompleteness rather than a deliberate decision.

**Comment by equwal** [comment]
> A quick Googling indicates that the lack of an xcb equivalent to XLookupString _seems_ to be a case of feature-incompleteness rather than a deliberate decision.

Yeah it gets much worse than that very fast. I think that CLX should throw in the towel and use some C FFI bindings rather than trying to reimplement everything in Common Lisp.

**Comment by dkochmanski** [comment]
the whole point of clx is that it doesn't depend on any kind of ffi. If you are interesting in having xlib wrapper then you may try to contribute in this project https://github.com/rpav/cl-xcb-xlib (it seems to not be maintained).

**Comment by equwal** [comment]
> the whole point of clx is that it doesn't depend on any kind of ffi. If you are interesting in having xlib wrapper then you may try to contribute in this project https://github.com/rpav/cl-xcb-xlib (it seems to not be maintained).

Thank you for that @dkochmanski.

Would an FFI for _new features only_ be acceptable, maybe to this project? https://xkbcommon.org/

I didn't intend to suggest replacing all of CLX with FFI bindings, just adding a few for this feature, since such an effort is likely to be more successful than implementing the xkeyboard extension in Common Lisp.

**Comment by dkochmanski** [comment]
I'm not sure whether it is even possible to shoehorn xlib calls in between "native" clx calls. If it is possible, then there is nothing what would prevent someone from creating an extension in a separate repository (and as a separate system), clx api won't change anytime soon (and it is even unlikely that internals will change except for fixed bugs), such system would simply depend-on clx.
